designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,300 @@
+
+### Goals
+
+* Kubebuilder is able to discover plugin binaries and run those plugins using the CLI.
+
+* Kubebuilder can use the external plugins as its own internal ones to do scaffolding.
+
+* Kubebuilder is able to discover flags that each individual plugin would support.
+
+* Kubebuilder should be able to show plugin specific information via the `--help` flag.
+
+* Support for stdin/stdout as the only IPC method between kubebuilder and plugins.
+
+* Operator SDK is able to extend Kubebuilder and provide the same functionality for plugins chaining and enabling out-of-tree plugin discovery.
+
+### Non-Goals
+
+* Addition of new arbitrary subcommands other than the subcommands that we already support i.e init, create api, and create webhook.
+
+* Discovering plugin binaries that are not locally present on the machine (i.e. binary exists in a remote repository).
+
+* Providing other options (other than stdin/stdout) for inter-process communication between Kubebuilder and plugins.
+
+### Examples
+
+* `kubebuilder create api --plugins=myexternalplugin/v1`
+  * should scaffold files using the external plugin as defined in its implementation of the create api method.
+
+* `kubebuilder create api --plugins=myexternalplugin/v1,myotherexternalplugin/v2`
+  * should scaffold files using the external plugin as defined in their implementation of the create api method (by respecting the plugin chaining order, i.e. in the order of create api of v1 and then create api of v2 as specified in the layout field in the configuration).
+
+* `kubebuilder create api --plugins=myexternalplugin/v1 --help`
+  * should provide help implementation of the plugin which is not shipped in the binary (plugin is present outside of the kubebuilder binary).
+
+## Proposal
+
+### Discovery of plugin binaries
+
+Main goal of Phase 2 plugins is discovering external plugins and running those plugins via the Kubebuilder CLI. I will describe an approach here that seems to be most suitable as it's a common way to configure plugins. In the Alternatives section below, I have discussed few other approaches that can be considered with their respective pros and cons.
+
+#### Approach #1
+
+This is similar to how [kustomize](https://kubectl.docs.kubernetes.io/guides/extending_kustomize/) discovers plugins by following a GVK path scheme.
+
+Every plugin gets its own directory constructed using the plugin name and plugin version for the executable to be placed in and Kubebuilder will search for a plugin binary with the name of the plugin in the `${name}/${version}` directory of the plugin. This information (plugin name and plugin version) is obtained to Kubebuilder via the value passed to the `plugins` field through the CLI. Once KB successfully locates the plugin, it will run the plugin using the CLI.
+
+Every plugin gets its own directory such as:
+
+```project
+    $XDG_CONFIG_HOME/kubebuilder/plugins
+    /${name}/${version}
```

Comment:
I had to specify a language as it's a fenced code block and leaving it empty was causing markdown lint errors. Switched to `shell` instead of `project` as it may fit better.